src/EditorFeatures/Core.Wpf/InlineDiagnostics/InlineDiagnosticsAdornmentManager.cs

```diff
@@ -207,5 +207,37 @@ protected override void AddAdornmentsToAdornmentLayer_CallOnlyOnUIThread(Normali
                     removedCallback: delegate { graphicsResult.Dispose(); });
             }
         }
+
+        protected override void UpdateSpans_CallOnlyOnUIThread(NormalizedSnapshotSpanCollection changedSpanCollection, bool removeOldTags)
+        {
+            Contract.ThrowIfNull(changedSpanCollection);
+
+            // this method should only run on UI thread as we do WPF here.
+            Contract.ThrowIfFalse(TextView.VisualElement.Dispatcher.CheckAccess());
+
+            var viewLines = TextView.TextViewLines;
+            if (viewLines == null || viewLines.Count == 0)
+            {
+                return; // nothing to draw on
+            }
+
+            // removing is a separate pass from adding so that new stuff is not removed.
+            if (removeOldTags)
+            {
+                foreach (var changedSpan in changedSpanCollection)
+                {
+                    // is there any effect on the view?
+                    if (viewLines.IntersectsBufferSpan(changedSpan))
+                    {
+                        // We add the entire extent of the line as the visual span when adding an adornment,
+                        // so we should remove based on that line.
+                        var line = viewLines.GetTextViewLineContainingBufferPosition(changedSpan.Start);
+                        AdornmentLayer.RemoveAdornmentsByVisualSpan(line.Extent);
```

Comment:
Does this now create a reverse problem where we'll remove the entire line's diagnostics but then only add the ones that touch the edit?